src/backend/renderer/damage.rs

```diff
@@ -0,0 +1,525 @@
+//! #         _: Size<i32, Buffer>,
+//! #         _: bool,
+//! #     ) -> Result<Self::TextureId, Self::Error> {
+//! #         unimplemented!()
+//! #     }
+//! #     fn update_memory(
+//! #         &mut self,
+//! #         _: &Self::TextureId,
+//! #         _: &[u8],
+//! #         _: Rectangle<i32, Buffer>,
+//! #     ) -> Result<(), Self::Error> {
+//! #         unimplemented!()
+//! #     }
+//! # }
+//! use smithay::{
+//!     backend::renderer::{
+//!         damage::DamageTrackedRenderer,
+//!         element::memory::{MemoryRenderBuffer, MemoryRenderBufferRenderElement}
+//!     },
+//!     utils::{Point, Transform},
+//! };
+//! use std::time::{Duration, Instant};
+//!
+//! const WIDTH: i32 = 10;
+//! const HEIGHT: i32 = 10;
+//! # let mut renderer = FakeRenderer;
+//! # let buffer_age = 0;
+//! # let log = slog::Logger::root(slog::Discard.fuse(), slog::o!());
+//!
+//! // Initialize a new damage tracked renderer
+//! let mut damage_tracked_renderer = DamageTrackedRenderer::new((800, 600), 1.0, Transform::Normal);
+//!
+//! // Initialize a buffer to render
+//! let mut memory_buffer = MemoryRenderBuffer::new((WIDTH, HEIGHT), 1, Transform::Normal, None);
+//!
+//! let mut last_update = Instant::now();
+//!
+//! loop {
+//!     let now = Instant::now();
+//!     if now.duration_since(last_update) >= Duration::from_secs(3) {
+//!         let mut render_context = memory_buffer.render();
+//!
+//!         render_context.draw(|_buffer| {
+//!             // Update the changed parts of the buffer
+//!
+//!             // Return the updated parts
+//!             vec![Rectangle::from_loc_and_size(Point::default(), (WIDTH, HEIGHT))]
+//!         });
+//!
+//!         last_update = now;
+//!     }
+//!
+//!     // Create a render element from the buffer
+//!     let location = Point::from((100.0, 100.0));
+//!     let render_element =
+//!         MemoryRenderBufferRenderElement::from_buffer(location, &memory_buffer, None, None);
+//!
+//!     // Render the output
+//!     damage_tracked_renderer
+//!         .render_output(
+//!             &mut renderer,
+//!             buffer_age,
+//!             &[render_element],
+//!             [0.8, 0.8, 0.9, 1.0],
+//!             &log,
+//!         )
+//!         .expect("failed to render the output");
+//! }
+//! ```
+
+use std::collections::VecDeque;
+
+use indexmap::IndexMap;
+
+use crate::{
+    backend::renderer::Frame,
+    output::Output,
+    utils::{Physical, Rectangle, Scale, Size, Transform},
+};
+
+use super::{
+    element::{Id, RenderElement},
+    utils::CommitCounter,
+};
+
+use super::{Renderer, Texture};
+
+#[derive(Debug, Clone, Copy)]
+struct ElementState {
+    last_commit: CommitCounter,
+    last_geometry: Rectangle<i32, Physical>,
+    last_z_index: usize,
+}
+
+#[derive(Debug, Default)]
+struct RendererState {
+    size: Option<Size<i32, Physical>>,
+    elements: IndexMap<Id, ElementState>,
+    old_damage: VecDeque<Vec<Rectangle<i32, Physical>>>,
```

Comment:
Could/should this be a Vec and use push() and drain() instead of push_front() and truncate()?

Reply:
If this is used like a ring buffer of sorts, a VecDeque may be more appropriate